### PR TITLE
Api key per client

### PR DESCRIPTION
**Pull request by joelsonoda**
This is how I preferred to configure the scamandrill client for my use case (one key per client vs one key per call (by setting in the model) or per application context (set in the application config).

I'm not sure if this is how other people would like to use it, but I thought I would share this.

Mostly, I wish that the API key were not part of the mandrill API request object but rather it were an Authorization header (and also that they used other verbs like GET...).

### CODE REVIEW

**Comment by graingert** [comment]
I'd like it to be obvious how to maintain the http pool across instances.

val s = new Scamandrill()
val client = s with APIKey("xxxxxxx")
val otherClient = client with APIKey("yyyyy")

Also I want to get rid of the Abstract client. It's just a long list of Any methods

**Comment by joelsonoda** [comment]
I was wondering about that Abstract client trait...it made more sense with the previous version, I supposed.

Is a Scamandrill a client or is it a client provider? That is, are the API methods exposed through `s` or does `s` only allow the creation of clients?

If it IS a client, we could probably leave it at MandrillClient and create a `with` method that yields another MandrillClient using the same HTTP(), right?

Thanks,
Joel

**Comment by graingert** [comment]
I see Scamandrill as a MandrillClient provider only.
MandrillClient could also be a MandrillClient provider

**Comment by graingert** [comment]
This would be v3 where I also want to switch to play-ws

**Comment by joelsonoda** [comment]
play-ws would make my application lighter since the rest of it is already play...let me know when you want to start and I'm glad to help. Meanwhile, I rescind this request as it is likely too large for v2.

**Comment by joelsonoda** [comment]
If scamandrill were play-ws based, we could just pass a WSClient to each MandrillClient and not really need the Scamandrill at all. 

Would you envision moving to play-json for the marshalling/unmarshalling as well?

**Comment by joelsonoda** [comment]
I've started putting together something using play-ws (only the very basic User requests so far).


If you get a chance, take a look at [this branch](https://github.com/joelsonoda/scamandrill/tree/api-key-per-client) again and let me know what your thoughts/ideas are on it. I tried to make it match your concept of Scamandrill as a provider for clients. Example usage is as follows.

```scala
import io.github.scamandrill.client._
// To make the future less boring
import scala.concurrent.ExecutionContext.Implicits.global

/**
  * This will construct Scamandrill in stand-alone mode with
  * the underlying WSClient created upon construction.
  * The project default settings in io.github.scamandrill.conf can be overridden
  * in the Mandrill section of application.conf (e.g. Mandrill.play.ws.followRedirects = true)
  */
val scamandrill = Scamandrill()
val client = scamandrill.getClient("API_KEY") // String or APIKey is fine
val clientTwo = client.getClient(APIKey("OTHER_KEY"))
val clientThree = clientTwo.getClient() // Using default key
val clientFour = cscamandrill.getClient() // Using default key

// Print a PONG!
scamandrill.getClient().usersPing.foreach(println)

/** 
  * This will only shutdown the client if the client is provided to Scamandrill
  * If Scamandrill constructs the WSClient, it will also shutdown the ActorSystem 
  * used to create that client
  */
scamandrill.shutdown() // or client(\w*).shutdown()
```

File: src/main/resources/io.github.scamandrill.conf [hunk 1]
```diff
@@ -0,0 +1 @@
+play.ws.followRedirects = true
```
Comment (graingert) [inline src/main/resources/io.github.scamandrill.conf]: missing new line

File: src/main/scala/io/github/scamandrill/client/MandrillClient.scala [hunk 2]
```diff
@@ -1,445 +1,1099 @@
...
+//  //MESSAGES calls https://mandrillapp.com/api/docs/messages.JSON.html
+//  ////////////////////////////////////////////////////////////////////
+//
+//  /**
+//    * Send a new transactional message through Mandrill
+//    *
+//    * @param msg - the message to send
+//    * @return - an of structs for each recipient containing the key "email" with the email address and "status" as either "sent", "queued", or "rejected"
+//    */
+//  def messagesSend(msg: MSendMessage): Future[List[MSendResponse]] = {
```
Comment (graingert) [inline src/main/scala/io/github/scamandrill/client/MandrillClient.scala]: commented code?

**Comment by graingert** [comment]
I'd like to get rid of the config entirely as well.

**Comment by joelsonoda** [comment]
Another idea to consider:
Modularize the client
```scala
class MandrillUserClient(...) extends ScamandrillSendReceive {
   def ping: Future[MandrillResponse[MPingResponse]] = {
       executeQuery[MVoid, MPingResponse](ping, MVoid())
   }
}
```

```scala
class MandrillClient extends ScamandrillSendReceive with MandrillClientProvider {
  val user = new MandrillUserClient(ws, key, ec)
}
```

```scala
client.user.ping()
```

File: src/main/scala/io/github/scamandrill/client/MandrillClient.scala (hunk 2 above)
Comment (joelsonoda) [inline src/main/scala/io/github/scamandrill/client/MandrillClient.scala]: Very much a work in progress.

I've only converted the very basic User commands from the API...everything else is commented out

**Comment by graingert** [comment]
Not sure about the modularization but it should probably be in another PR

**Comment by joelsonoda** [comment]
Thanks...I wasn't sure if commenting on a closed pull request would send notifications or not.
It's very much a work in progress at this point, but I wanted to check in with you and see if these kind of changes were in line with your vision for the API before converting the whole API.

**Comment by graingert** [comment]
also to override ws config I'd rather it was just passed in

**Comment by graingert** [comment]
you can also just use withFollowRedirects in a wrapper class

**Comment by joelsonoda** [comment]
Changed it to use a Configuration rather than using application.config to override defaults...

We don't actually need to change the `play.ws.followRedirects` since the default value is `true`. I just put that in there as an example in case there were other things about the client we wanted to use as project defaults. We probably can get away with using `Configuration.reference` for this project.

**Comment by joelsonoda** [comment]
I finished implementing all the calls with play-ws as well as filling in tests for each call.

File: src/main/resources/io.github.scamandrill.conf (hunk 1 above)
Comment (graingert) [inline src/main/resources/io.github.scamandrill.conf]: I thought this file should be removed

File: src/main/scala/io/github/scamandrill/client/ScamandrillSendReceive.scala [hunk 3]
```diff
@@ -1,70 +1,60 @@
...
-        }
-      }
+  def executeQuery[Req, Res](endpoint: Endpoint, model: Req)(implicit writes: Writes[Req], reads: Reads[Res]): Future[Try[Res]] = {
+    def handleError(res: WSResponse): Try[Res] = {
+      res.json.validate[MandrillError].fold(
+        invalid = _ => Failure(new UnsuccessfulResponseException(res)),
+        valid = me => Failure(new MandrillResponseException(res, me))
+      )
+    }
+    ws.url(s"$serviceRoot$endpoint").post(Json.toJson(model)(authenticatedWriter(writes))) map {
```
Comment (graingert) [inline src/main/scala/io/github/scamandrill/client/ScamandrillSendReceive.scala]: You can use:
```ws.url(s"$serviceRoot$endpoint").withFollowRedirects(true)```

**Comment by joelsonoda** [comment]
We've been running this branch in production for the last few weeks with good results. Since we're using playframework, it cleans up my dependencies nicely and I can use a guice to inject scamandrill into whichever classes require it.

It might be nice to add a guice module for scamandrill before 3.0 so that the playframework folks can just include the module in their application.conf and not need to implement their own provider.

Another nice-to-have would be changing the image/attachment class to take either a String or Source[ByteString, Any] and the mandrill client could see to converting to a base64 string.

I have my own wrapper around that class in my applications, but I think pretty much everyone would need to implement the same thing in order to include images and send attachments.

File: src/main/resources/io.github.scamandrill.conf (hunk 1 above)
Comment (graingert) [inline src/main/resources/io.github.scamandrill.conf]: This file should still be removed

**Comment by graingert** [comment]
Looks great, can you clean up your commits

File: build.sbt [hunk 4]
```diff
@@ -25,18 +26,14 @@ resolvers ++= Seq("spray repo" at "http://repo.spray.io/")
 parallelExecution in Test := true
```
Comment (graingert) [inline build.sbt]: You don't need the spray repo anymore

**Comment by joelsonoda** [comment]
Also, I removed most of the tests that actually hit the API since they won't work unless one uses a specific mandrill account. We could add some like that back in (I have some examples for the user ping tests), and then the developer can opt not to run those tests or to run only those tests (they are skipped if no SCAMANDRILL_API_KEY is present).

File: src/main/scala/io/github/scamandrill/client/MandrillClientProvider.scala [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  val onShutdown: () => Future[Unit]
+
+  /**
+    * Closes the underlying WSClient.
+    * Users of this class are supposed to call
+    * this method when they are finished using Scamandrill.
+    */
+  def shutdown(): Future[Unit] = {
+    ws.close()
+    onShutdown()
```
Comment (graingert) [inline src/main/scala/io/github/scamandrill/client/MandrillClientProvider.scala]: I don't see why onShutdown is useful

**Comment by graingert** [comment]
@joelsonoda yeah I'd like the integration tests, as I have those configured with a private travis variable.

If you start work on them I'll add you as a contributor

File: src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package io.github.scamandrill.models
+
+import io.github.scamandrill.MandrillSpec
+
+class MandrillMessagesRequestsTest extends MandrillSpec {
+  "MSendMsg" should "copy to an equal msendmessage" in {
+    val msg = new MSendMsg(
+      html = "<h1>HTML CONTENT</h1",
```
Comment (graingert) [inline src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala]:
```scala
<h1>HTML CONTENT</h1>.mkString
```

File: src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      subject = "I'm a Subject",
+      from_email = "bmcboatface@nerc.ac.uk",
+      from_name = "Boaty McBoatface",
+      to = List(MTo(
+        email = "test@test.com",
+        name = Some("Test Testerson")
+      ))
+    )
+
+    msg shouldBe msg.copy()
```
Comment (graingert) [inline src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala]:
```scala
msg.copy() ==== msg
```

File: src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package io.github.scamandrill.models
+
+import io.github.scamandrill.MandrillSpec
+
+class MandrillMessagesRequestsTest extends MandrillSpec {
+  "MSendMsg" should "copy to an equal msendmessage" in {
+    val msg = new MSendMsg(
```
Comment (graingert) [inline src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala]: Isn't MSendMsg a case class? (use with apply)

File: src/test/resources/responses/webhooks/info.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+        "send",
+        "open",
+        "click"
+    ],
+    "created_at": "2013-01-01 15:30:27",
+    "last_sent_at": "2013-01-01 15:30:49",
+    "batches_sent": 42,
+    "events_sent": 42,
+    "last_error": "example last_error"
+}
```
Comment (graingert) [inline src/test/resources/responses/webhooks/info.json]: missing trailing new line. Please fix these everywhere

File: src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala (hunk 8 above)
Comment (joelsonoda) [inline src/test/scala/io/github/scamandrill/models/MandrillMessagesRequestsTest.scala]: Alas, there are too many properties for an apply method on MSendMsg.

**Comment by joelsonoda** [comment]
I added back in the actual tests for inbound domains and ip calls.

**Comment by graingert** [comment]
great can you clean up the commits, eg so instead of the tests being added and removed it's as though they were never removed in the first place. Same with the extra config file.

**Comment by joelsonoda** [comment]
I'll finish porting back the other tests and then rewrite my commits.

**Comment by graingert** [comment]
Great

**Comment by joelsonoda** [comment]
I'll be back with a new pull request once those are done

**Comment by graingert** [comment]
@joelsonoda you can just force push to this same branch and keep this PR open

**Pull request closed without merging**